src/foam/nanos/ruler/RulerDAO.js

```diff
@@ -168,12 +170,44 @@ for ( Object key : sink.getGroupKeys() ) {
     ((foam.nanos.logger.Logger) x.get("logger")).error("RuleGroup not found.", key);
   } else if ( rg.f(x, obj, oldObj) ) {
     List<Rule> group = ((ArraySink) sink.getGroups().get(key)).getArray();
-    if ( ! group.isEmpty() ) {
-      new RuleEngine(x, getX(), this).execute(group, obj, oldObj);
+    var rules = enforceSpid(x, obj, group);
+    if ( ! rules.isEmpty() ) {
+      new RuleEngine(x, getX(), this).execute(rules, obj, oldObj);
     }
   }
 }`
     },
+    {
+      name: 'enforceSpid',
+      type: 'List<Rule>',
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'obj', type: 'FObject' },
+        { name: 'rules', type: 'List<Rule>' }
+      ],
+      javaCode: `
+        return rules.stream()
+          .filter(rule -> rule.isGlobalSpid() || isSpidMatched(x, rule, obj))
+          .collect(Collectors.toList());
+      `
+    },
+    {
+      name: 'isSpidMatched',
+      type: 'boolean',
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'rule', type: 'Rule' },
+        { name: 'obj', type: 'FObject' }
+      ],
+      documentation: 'Send a ServiceProviderAwareSupport action command to the DAO for spid matching test.',
+      javaCode: `
+        var result = getDelegate().cmd_(x.put("OBJ", obj),
+          new ServiceProviderAwareSupport.Builder(x)
```

Comment:
Create an explicit constructor for this so we don't have the garbage from using the builder.  Thanks.

Reply:
Fixed in https://github.com/foam-framework/foam2/pull/3680/commits/8ab51fc22232bd26c9e40abf4b35adebf3675232.